### PR TITLE
Alternative Entity type casting

### PR DESCRIPTION
**Pull request by iRedds**
**Description**
I propose an alternative implementation for type casting. 

- All type casts that existed by default have been moved to separate classes. 
- ~~Reworked magic methods __get and __set~~.
- ~~The setter method now has call priority.~~
- ~~If a getter/setter and type castings are defined, then only the getter/setter will be called.~~
- Added functionality for defining custom type casting. Custom type handlers have get and set methods. 
- Types can now have additional parameters.
- ~~All default type casts cannot be overridden.~~

Example
```php
    // Custom type registration 
    class MyEntity extends Entity
    {
        // Specifying the type for the field
        protected $casts = [
            'key' => 'base64',
        ];

        //Bind the type to the handler
        protected $castHandlers = [
            'base64' => 'App\Entity\Cast\CastBase64',
        ];
    }

    // Handler 
    class CastBase64 extends \CodeIgniter\EntityCast\AbstractCast
    {
        public static function get($value, array $params = [])
        {
            return base64_decode($value);
        }

        public static function set($value, array $params = [])
        {
            return base64_encode($value);
        }
    }
```

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by michalsn** [comment]
I like this. It adds a lot of flexibility but I think we have to make some of these features optional:

> * The setter method now has call priority.
> * If a getter/setter and type castings are defined, then only the getter/setter will be called.

If I'm not mistaken, those two will break the compatibility with the current version. So we have two options here: we can either get back to what we have now or make this an optional feature.

I can imagine that we can set a special flag in Entity that will make it work, so we can choose the type of functionality we expect from our entity. I know that it adds some complexity, but backward compatibility for Entities is a must-have.

File: system/Entity.php
Comment (najdanovicivan) [inline system/Entity.php:532]: Should be inverted so that castHandles can override defaultCastHandlers

**Review by najdanovicivan — COMMENTED** [review]
That's great.  Is there an ETA when this can be merged ?

Additionally I'd add property $initCastHandlers = [] and $defaultInitCastHandlers which would trigger casting in setAttributes method.

Then I'd introduce Config\Entity where defaultInitCastHandlers and defaultCastHandlers can be defined. Preferable it should be possible to Override this per module.

Another thing I'd like is to be able to define those casts to work with reference values. For example I have some item entities which have categories. And I want to display a lot of those on single page. I have slug for category in item. Then I load all the categories in some array. What I do next is only set a pointer to the category object inside item instead of copying those into items. That way memory usage can be reduced by a bit.

Also note the issue below

File: system/Entity.php
Comment (iRedds) [reply system/Entity.php]: I had doubts about replacing the default handlers.
Your opinion helped me decide

**Comment by iRedds** [comment]
@michalsn My point of view is this. Setters and Getters are needed for special cases when you need to bypass type casting and implement other logic. 
I believed the implementation was the wrong implementation at the time of design. Therefore, I took responsibility for making changes in the behavior of Getters and Setters. 
At the moment, I don't know how best to implement the ability to switch to the priority mode of Getters and Setters without the headache, duplicate and ugly code.
Maybe someone will succeed.
But now I decided to return the old behavior.

**Comment by iRedds** [comment]
@najdanovicivan I think the merge will be possible when the maintainers have no questions about the current implementation.

Can't imagine what the cast is for in the setAttributes method. 
Perhaps you mean some special case. 

If you like my PR, I think it would be better to wait for it to merge. And then you can submit your implementation based on my changes. It will be more clear with the code and examples.

**Comment by najdanovicivan** [comment]
@iRedds cast in setAttributes is useful for nosql when you have nested objects. Let's take this one for example. We have a ecommerce site which support product variations which different pricing options. So you have something like this in MongoDB


```
{
    "varations" [
          {
              "name" : "item-only",
              "price": 1000
          },
          {
              "name" : "item-with-case",
              "price": 1500
          }
          {
              "name" : "item-with-accessories",
              "price": 2000
          }
     ]
}
```

We have 2 classes. Product and ProductVariation so we need to be able to add/modify those variations

```
$variation = new ProductVariation()
$variation->name = 'new'
$variation->price = 3000

//This one is not possible with magic for the time being
$product->variations[] = $variation;

//So we have to do
$variations = $product->variations; 
$variations[] = $variation;
$product->variations = $variations;

//This will work with setter/getter casting but will do the object construction each time getter is called

//Now this one will do completely nothing on the entity as setter will never be triggered
$product->variations[0]->name = 'New name';
```

Regarding the DB reading/writing we have toRawArray method for which if $deep parameter is passed as true it will handle the conversion of inner entities and we'll get the complete data to write in the db.

Problem with setAttributes is that it never goes through the setters so we'll have array in attributes instead of having an entity object stored in attributes.

**Comment by iRedds** [comment]
@najdanovicivan Now I understand what you were talking about.
I have a persistent feeling that earlier on CI3 using Entity I faced a similar problem, but I cannot find where.
As far as I understand the concept of Entity applied during development is that the data should exist "AS IS". That is, having received a string from the database, we only transform it to receive it, but the data itself does not change.

I don't think it would be correct to change the behavior of the setAttributes method.
What is now done in the toRawArray method is a crutch, which, moreover, does not work correctly. 
Another approach must be found.

**Comment by najdanovicivan** [comment]
https://github.com/codeigniter4/CodeIgniter4/blob/4cbfe2cd25ff8b08eee323cc734a2113cead8eb1/system/Database/MySQLi/Result.php#L159-L175

Looking at these the only thing setting the values inside entity form db is the setAttributes method. 

By adding defaultInitCastHandlers/InitCastHandlers we can allow for such behavior. Those will be empty by default so the existing functionality won't be affected but it will allow the feature to be used if needed. 

Having defaultInitCastHandlers in Config/Entity just provide means to setup those handlers globally. In the same manner it's done with validation rules. So that there is no need to define those in every Entity class

File: system/Entity.php
Comment (najdanovicivan) [inline system/Entity.php:556]: We should have `nullable` added in parameters `if (strpos($type, '?') === 0)`

There might be a situation where `null` might have to be return instead of $value. For example casting empty string to `?int`. The cast method should return null in such case but having `int` as cast should trow an exception or maybe in some custom cast class return 0.

**Review by najdanovicivan — CHANGES_REQUESTED** [review]

File: system/Entity.php
Comment (iRedds) [reply system/Entity.php:556]: The nullable type is handled above.
I see no reason to apply nullable to parameters. They are not computed and are passed to the handler as a string

**Comment by iRedds** [comment]
I understand the reason why you want to change the setAttributes method. But at the moment I am what the correct implementation should be. And I think that it is outside the scope of the current PR. 

There are 2 places in the setAttributes method where you can convert. 
1. Before $this->attributes = $data;
2. After $this->syncOriginal(); 

In the first case, when converting to an object, we will receive the same object passed by reference in the $attributes and $original arrays. And when the object changes, it will not be processed if only the changed attributes need to be retrieved. 
```php 
$a = ['a' => new stdClass];
$b = $a;
$b['a']->a = 1;
var_dump($a === $b); // true
```
In the second case, the attribute will be in the "changed" status all the time, even if it has not changed.
Moreover, the toRawArray (true, true) method does not work correctly.

Do you have a solution for this?

File: system/Entity.php
Comment (najdanovicivan) [reply system/Entity.php:556]: There are situations where it might be useful. There is no way to submit null via form. So in most cases I have custom setters which will set the empty string got via post to null. 

Having nullable as parameter in custom setter object will make it possible to completely get rid of such setters as it will be possible to make custom cast which will handle empty string to null conversion.

**Comment by najdanovicivan** [comment]
We can modify hasChanged() method to iterate attributes check if value is of Entity class and if so it will forward the call to the inner Entity object 

I already did something very similar for to array method with introduction of the $recursive parameter. 

Then it makes more sense of doing conversion in `setAttributes()` instead of `syncOriginal()`

File: system/Entity.php
Comment (iRedds) [reply system/Entity.php:556]: How do you imagine that? I don't understand.
'field' => 'my_type[?param]'  Like this?

Comment (najdanovicivan) [reply system/Entity.php:556]: set $isNullable inside if (strpos($type, '?') === 0)

And then here after pregmatch append it to the params

if($isNullable)
{
  $params[] = ‘nullable’;
}

**Comment by najdanovicivan** [comment]
I've given some thought into this and got an interesting idea. 

@iRedds what do you thing about extending the CastInterface with 2 additional methods

```
```

File: system/Entity.php
Comment (najdanovicivan) [reply system/Entity.php:556]: That's a good question. It can be done that way but it makes things very confusing as there will be 3 cases which allow null values.

`?bool`
`?bool[nullable]`
`bool[nullable]`

**Comment by iRedds** [comment]
At the moment the question is whether this PR will be merged )

File: system/Entity.php
Comment (iRedds) [reply system/Entity.php:560]: OK. Commit with your suggestion added

File: system/Entity.php
Comment (paulbalandan) [inline system/Entity.php:518]: Add a default value to `$method` so it won't break user's implementations using this method.

File: system/Entity.php
Comment (paulbalandan) [inline system/Entity.php:558]:
```suggestion
			$params = array_map('trim', explode(',', $matches[2]));
```

File: system/EntityCast/AbstractCast.php
Comment (paulbalandan) [inline system/EntityCast/AbstractCast.php:9]: Remove the package tags

File: system/EntityCast/AbstractCast.php
Comment (paulbalandan) [inline system/EntityCast/AbstractCast.php:2]: Add the license headers

File: system/EntityCast/CastAsArray.php
Comment (paulbalandan) [inline system/EntityCast/CastAsArray.php:18]: Do not add space before the colon.
```suggestion
	public static function get($value, array $params = []): array
```

File: system/EntityCast/CastAsDouble.php
Comment (paulbalandan) [inline system/EntityCast/CastAsDouble.php:12]: This should be the same with `CastAsFloat` so no longer necessary.

File: system/EntityCast/CastAsJson.php
Comment (paulbalandan) [inline system/EntityCast/CastAsJson.php:7]: Alphabetically order these two.

File: system/EntityCast/CastAsJson.php
Comment (paulbalandan) [inline system/EntityCast/CastAsJson.php:31]: You can use here the `JSON_THROW_ON_ERROR` flag and catch any `JsonException`

File: system/EntityCast/CastAsJson.php
Comment (paulbalandan) [inline system/EntityCast/CastAsJson.php:48]: Use `JSON_THROW_ON_ERROR`

File: system/EntityCast/CastAsTimestamp.php
Comment (paulbalandan) [inline system/EntityCast/CastAsTimestamp.php:20]: Maybe you can additionally check the return value if not `false`. If `false`, then throw an error for invalid timestamp.

File: user_guide_src/source/models/entities.rst
Comment (paulbalandan) [inline user_guide_src/source/models/entities.rst:521]: Indent

Comment (iRedds) [reply user_guide_src/source/models/entities.rst]: I try to follow the PSR-12 style, for example 120 characters per line.

Comment (paulbalandan) [reply user_guide_src/source/models/entities.rst]: Ah. I mean the `the value nullable...`. This causes parsing errors in sphinx.

**Comment by iRedds** [comment]
* License comment added to all files in PR .
* In the Entity :: castAs () method, the $ method parameter is set to its default value. 
* Removed @package in docBlock.
* JSON casting now throws an exception instead of json_last_error(). 
* Timestamp casting now throws an exception if the value cannot be converted to a number.
* Double casting is now an alias for float.

File: user_guide_src/source/models/entities.rst
Comment (paulbalandan) [inline user_guide_src/source/models/entities.rst:521]: This needs to be also indented. Sphinx warns `Explicit markup ends without a blank line; unexpected unindent.`. Align this with `note` above.

File: system/Exceptions/CastException.php
Comment (paulbalandan) [inline system/Exceptions/CastException.php:60]: Framework's way of declaring new static methods is `for` + the statement. So in this case it would be `forInvalidTimestamp` etc,

File: user_guide_src/source/models/entities.rst
Comment (iRedds) [reply user_guide_src/source/models/entities.rst:521]: I'm sorry. I was stupid.

File: system/Exceptions/CastException.php
Comment (paulbalandan) [reply system/Exceptions/CastException.php]: I'm sorry I wasn't clear enough. What I meant is that all new methods would begin with `for`, not only `invalidTimestamp`.

Comment (iRedds) [reply system/Exceptions/CastException.php]: Fixed

**Review by paulbalandan — APPROVED** [review]
Last thing, please add an entry to the changelog in `user_guide_src`.

**Review by najdanovicivan — APPROVED** [review]

**Review by michalsn — APPROVED** [review]

**Comment by paulbalandan** [comment]
Thank you, @iRedds

**Comment by iRedds** [comment]
Thank you all for the help with PR

**Pull request merged**
